### PR TITLE
Flink: Fix tests creating catalog after FLINK-29677

### PR DESCRIPTION
**Pull request by pvary**
FLINK-29677 add a check, so the currently used `Catalog` could not be dropped.
FLINK-29677 will land in Flink 1.16.1, and 1.17.0 and when we start to use it we need to update some of our tests.

This PR creates a method `pushCatalog` to store the current catalog, and `popCatalog` to restore this old catalog. This should be used to restore the original catalog before dropping the one used in the tests.

### CODE REVIEW

**Review by nastra — COMMENTED** [review]
rather than using `pushCatalog()`/`popCatalog()` wdyt about having a `dropCatalog(<name>, <if exists handling>)` helper method that switches to the `default_catalog` and then executes the `DROP CATALOG <if exists> <name>` stmt?

**Comment by stevenzwu** [comment]
> switches to the default_catalog and then executes the DROP CATALOG <if exists> <name>

I like @nastra 's suggestion. also maybe add some code comment to explain why we need to switch the catalog before calling `DROP CATALOG ...`

File: flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/FlinkTestBase.java [hunk 1]
```diff
@@ -113,4 +113,15 @@ protected void assertSameElements(String message, Iterable<Row> expected, Iterab
...
         .containsExactlyInAnyOrderElementsOf(expected);
   }
+
+  /**
+   * We can not drop currently used catalog after FLINK-29677, so we have make sure that we do not use the current
+   * catalog before dropping it. This method switches to the 'default_catalog' and drops the one requested.
+   * @param catalogName The catalog to drop
+   * @param ifExist If we should use the 'IF EXISTS' when dropping the catalog
+   */
+  protected void dropCatalog(String catalogName, boolean ifExist) {
```
Comment (nastra) [inline flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/FlinkTestBase.java:123]: nit: ifExist -> ifExists

File: flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/source/TestFlinkTableSource.java [hunk 2]
```diff
@@ -97,6 +97,7 @@ public void before() {
   public void clean() {
     sql("DROP TABLE IF EXISTS %s.%s", DATABASE_NAME, TABLE_NAME);
     sql("DROP DATABASE IF EXISTS %s", DATABASE_NAME);
+    dropCatalog(CATALOG_NAME, true);
     sql("DROP CATALOG IF EXISTS %s", CATALOG_NAME);
```
Comment (nastra) [inline flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/source/TestFlinkTableSource.java:101]: I think this line needs to be removed, given that we call dropCatalog above

**Review by nastra — APPROVED** [review]

File: flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/source/TestFlinkTableSource.java (hunk 2 above)
Comment (pvary) [reply flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/source/TestFlinkTableSource.java]: Thanks for catching this

File: flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/FlinkTestBase.java (hunk 1 above)
Comment (pvary) [reply flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/FlinkTestBase.java]: Fixed

**Review by stevenzwu — APPROVED** [review]

**Comment by stevenzwu** [comment]
thanks @pvary for the contribution and @nastra for reviewing.

**Pull request merged**